### PR TITLE
Store configs per session in the backend

### PR DESCRIPTION
**Pull request by aliabid94**
We now allow sessions to have different configs. This PR:
1. Creates a BlocksConfig object. This contains 
  a) BlocksConfig.blocks: a dictionary record of all Block objects for this session, and 
  b) BlocksConfig.fns: a list of all BlockFunction event listeners for this session. 
A Blocks object has a default BlocksConfig object
2. Each new session gets a new BlocksConfig by copying the default BlocksConfig, which is copied via a shallow copy of (a) and (b).
3. Component updates modify (a) of the BlocksConfig for this session. We simply create a new Component on an update and point to this new Component in (a) for this BlocksConfig
4. Previously, we stored event listener information across two data structures: blocks.fns (list of BlockFunctions) and blocks.dependencies (list of dicts). Have refactored to remove blocks.dependencies entirely. Removes duplication and concerns of keeping data in sync between two data structures, as well as eliminates the untyped code involving `blocks.dependencies`

### CODE REVIEW

File: gradio/blocks.py [hunk 1]
```diff
@@ -718,6 +825,22 @@ def __init__(
 
         self.queue()
 
+    @property
+    def blocks(self) -> dict[int, Component | Block]:
+        return self.default_config.blocks
+
+    @blocks.setter
+    def name(self, value: dict[int, Component | Block]):
```
Comment (abidlabs) [inline gradio/blocks.py:833]:
```suggestion
    def blocks(self, value: dict[int, Component | Block]):
```

File: gradio/helpers.py [hunk 2]
```diff
@@ -294,6 +294,7 @@ async def load_example(example_id):
                 api_name=self.api_name,
                 show_api=False,
             )
+            self.load_input_event_id = len(Context.root_block.fns) - 1
```
Comment (abidlabs) [inline gradio/helpers.py:297]: What's this?

**Comment by abidlabs** [comment]
The block function-related code is a lot better now, thanks @aliabid94 for making these changes! I tested a variety of demos and they all worked well. Noticed an semi-related bug, so I pushed a fix here (https://github.com/gradio-app/gradio/pull/8030/commits/02d56615d0c3ea2ba7ecba19f2f951979bcbfd04).

Left a couple of suggestions, but otherwise LGTM

**Review by abidlabs — APPROVED** [review]

File: gradio/helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliabid94) [reply gradio/helpers.py:297]: I think it's possible that that when we call `Context.root_block.fns.pop` later in `Examples.cache()`, there may have been more functions created between `Examples.create()` where we create the load event, and `Examples.cache()` where we pop it. There is a `_defer_cache` that I believes allows this to happen.

[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/helpers.py:297]: Ah yeah good point

**Pull request merged**